kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/ResourcesUtil.java

```
@@ -224,4 +241,168 @@ static <O extends HasMetadata, R extends HasMetadata> Set<ResourceID> findReferr
                 owner,
                 primary -> ResourcesUtil.isReferent(refAccessor.apply(primary), referent));
     }
+
+    static List<Condition> maybeAddOrUpdateCondition(List<Condition> conditions, Condition condition) {
+        var type = Objects.requireNonNull(condition.getType());
+
+        Comparator<Condition> conditionComparator = Comparator
+                .comparing(Condition::getType)
+                .thenComparing(Comparator.comparing(Condition::getObservedGeneration).reversed())
+                .thenComparing(Condition::getStatus)
+                .thenComparing(Condition::getReason)
+                .thenComparing(Condition::getMessage)
+                .thenComparing(Condition::getLastTransitionTime);
+
+        TreeMap<Condition.Type, List<Condition>> byType = conditions.stream().collect(Collectors.groupingBy(
+                Condition::getType,
+                TreeMap::new, // order based on Type
+                Collectors.toList()));
+
+        var map = byType.entrySet().stream().map(entry -> {
+            // minimum must exist because values of groupingBy result are always non-empty
+            Condition minimumCondition = entry.getValue().stream().min(conditionComparator).orElseThrow();
+            var mutableList = new ArrayList<Condition>(1);
+            mutableList.add(minimumCondition);
+            return Map.entry(entry.getKey(), mutableList);
+        }).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue,
+                (l1, l2) -> {
+                    throw new IllegalStateException();
+                },
+                TreeMap::new));
+
+        ArrayList<Condition> conditionsOfType = map.computeIfAbsent(type, k -> new ArrayList<>());
+        if (conditionsOfType.isEmpty()) {
+            conditionsOfType.add(condition);
+        }
+        else if (conditionsOfType.get(0).getObservedGeneration() <= condition.getObservedGeneration()) {
+            conditionsOfType.set(0, condition);
+        }
```

This means that if our condition has a generation that is older than the current highest it gets left on the floor. Should we sort it into position in the list? Probably not as it sort of implies we are acting on stale data so maybe we should the throwing an exception in that case to abort the reconciliation pass?

---

This shouldn't happen. It would be unambiguously incorrect to replace a condition with a higher generation with the given one. So either we drop the given condition, we grow the list, or we throw an exception.

* I don't think it helps to keep multiple conditions of the same type around. It gives no useful or actionable information to the user (let's not forget that `status` is our API to a user). We'd still need some logic to prune them, and that's gonna be difficult to test precisely because this should never happen.
* I don't think throwing is ideal either. That will manifest as a weird exception in the `status`, which the user likely won't understand or know how to fix. They might not even be _able_ to fix the problem. And whatever the cause is, it's very likely not the operator that's at fault.

---

>I don't think throwing is ideal either. That will manifest as a weird exception in the status, which the user likely won't understand or know how to fix. They might not even be able to fix the problem. And whatever the cause is, it's very likely not the operator that's at fault.

I don't think it should manifest in the status but I do think its a single that something has gone wrong and the reconciler should abort what ever it currently has in progress as its acting on stale data (isn't that the implication?). 

I think it means in the best case the operator is acting on stale data, at worst doesn't it imply that the operator is screwing something up by messing with generations? OHh could it be sign that there are multiple copies of the operator running (I don't think thats something we intend to support)

---

> I think it means in the best case the operator is acting on stale data, at worst doesn't it imply that the operator is screwing something up by messing with generations? OHh could it be sign that there are multiple copies of the operator running (I don't think thats something we intend to support)

It cannot occur due to stale data because the apiserver updates to a resource should be atomic and cover the whole resource. Whatever it later serves up (via a GET or a WATCH) will have a consistent `metadata`, `spec` and `status`. And the operator is updating the `status` based on the `spec` in the same request.

_If_ it happened it would imply that _something_ was messing with the `status.generation` (i.e. creating a `status` with a `status.observedGeneration` that was greater than the `metadata.generation`). The problem is we can't assume that that something is _this_ (or _any_) instance of the operator. 

Assume for a moment that you're right and there is some errant process somewhere. Suppose it eventually gets killed. What is this instance of the operator going to do _then_? What you're arguing for is for it to continue throwing an exception, because it will still be true that the `observedGeneration` on the `status` is greater than expected. I.e. the operator won't self-heal. Someone will have to fix it, if they can figure out what they need to do from the exception. It seems that `kubectl` now supports editing the `status`, so I suppose fixing it is easier that it once was, but it's that's a relatively recent addition. I really don't think that throwing in this situation is making anyone's life better.

---

but this change detects that we did a bunch of work and that something is very screwy  and silently ignores that fact. Is it safe to assume that any changes we are trying to apply are actually safe or sensible to apply if something else is messing with the CR?